src/sidebar/services/test/frame-sync-test.js

```diff
@@ -96,16 +89,34 @@ describe('FrameSyncService', () => {
     };
 
     fakeStore = fakeReduxStore(
-      { annotations: [] },
+      { annotations: [], frames: [] },
       {
         allAnnotations() {
           return this.getState().annotations;
         },
-        connectFrame: sinon.stub(),
-        destroyFrame: sinon.stub(),
```

Comment:
Instead of `frames`, `connectFrame` and `destroyFrame` all being independent mocks, they now functions that operate on the same Redux state. This ensures that the result of `frames()` is consistent with any prior `connectFrame` and `destroyFrame` calls.

Reply:
> A philosophical comment: I am a little bit divided about the mocking dependencies that are so intimately coupled, as for example the redux store and FrameSyncService.
> ...
> I know that integration tests, as we understand them in our group, are mean to test these interdependencies, but I sometimes wonder if we are mocking too much.

I think this is a very reasonable question to ask. I'm not a complete purist when it comes to unit vs integration testing. In this case I agree there are some risks to this more complex mocking. The downside of mocking complex dependencies are typically issues like:

- Tests running slowly
- Tests requiring a lot of effort to set up dependencies of dependencies
- Problems caused by side effects of dependencies
- Debugging test failures becoming more difficult because of the increased surface area they cover
- Negative impact on unit API design because there is less incentive to ensure that interfaces between modules are simple (and thus easy to mock)

An entirely pure Redux store (ours is "mostly pure") actually avoids many of these problems. If we removed the remaining side effects from the store it would be even easier.

Out of interest I looked into what it would be like if this test used the real store, but with various store methods stubbed. It turned out not to be too difficult.

```diff
diff --git a/src/sidebar/services/test/frame-sync-test.js b/src/sidebar/services/test/frame-sync-test.js
index 79882c75f..e7b53fcfd 100644
--- a/src/sidebar/services/test/frame-sync-test.js
+++ b/src/sidebar/services/test/frame-sync-test.js
@@ -6,6 +6,7 @@ import { fakeReduxStore } from '../../test/fake-redux-store';
 import { delay } from '../../../test-util/wait';
 
 import { FrameSyncService, $imports, formatAnnot } from '../frame-sync';
+import { createSidebarStore } from '../../store';
 
 class FakeWindow extends EventTarget {
   constructor() {
@@ -88,44 +89,16 @@ describe('FrameSyncService', () => {
       discover: sinon.stub().resolves(sidebarPort),
     };
 
-    fakeStore = fakeReduxStore(
-      { annotations: [], frames: [] },
-      {
-        allAnnotations() {
-          return this.getState().annotations;
-        },
-
-        connectFrame(frame) {
-          const frames = [
-            ...this.getState().frames,
-            {
-              ...frame,
-              isAnnotationFetchComplete: true,
-            },
-          ];
-          this.setState({ frames });
-        },
-
-        destroyFrame(frame) {
-          const frames = this.getState().frames;
-          this.setState({ frames: frames.filter(f => f.id !== frame.id) });
-        },
-
-        frames() {
-          return this.getState().frames;
-        },
-
-        findIDsForTags: sinon.stub(),
-        focusAnnotations: sinon.stub(),
-        isLoggedIn: sinon.stub().returns(false),
-        openSidebarPanel: sinon.stub(),
-        selectAnnotations: sinon.stub(),
-        selectTab: sinon.stub(),
-        setSidebarOpened: sinon.stub(),
-        toggleSelectedAnnotations: sinon.stub(),
-        updateAnchorStatus: sinon.stub(),
-      }
-    );
+    fakeStore = createSidebarStore({});
+    sinon.stub(fakeStore, 'findIDsForTags');
+    sinon.stub(fakeStore, 'focusAnnotations');
+    sinon.stub(fakeStore, 'isLoggedIn');
+    sinon.stub(fakeStore, 'openSidebarPanel');
+    sinon.stub(fakeStore, 'selectAnnotations');
+    sinon.stub(fakeStore, 'selectTab');
+    sinon.stub(fakeStore, 'setSidebarOpened');
+    sinon.stub(fakeStore, 'toggleSelectedAnnotations');
+    sinon.stub(fakeStore, 'updateAnchorStatus');
 
     fakeWindow = new FakeWindow();
     fakeWindow.parent = new FakeWindow();
@@ -214,9 +187,7 @@ describe('FrameSyncService', () => {
       await connectGuest();
       emitGuestEvent('documentInfoChanged', frameInfo);
 
-      fakeStore.setState({
-        annotations: [fixtures.ann],
-      });
+      fakeStore.addAnnotations([fixtures.ann]);
 
       assert.calledWithMatch(
         guestRPC().call,
@@ -228,12 +199,14 @@ describe('FrameSyncService', () => {
     it('sends annotations only to matching frame', async () => {
       // Annotation whose URI exactly matches the main frame.
       const mainFrameAnn = {
+        $tag: 't1',
         id: 'abc',
         uri: 'https://example.com',
       };
 
       // Annotation whose URI exactly matches the iframe.
       const iframeAnn = {
+        $tag: 't2',
         id: 'def',
         uri: 'https://example.com/iframe',
       };
@@ -243,6 +216,7 @@ describe('FrameSyncService', () => {
       // The search API can return such annotations if the annotation's URI
       // is deemed equivalent to one of the searched-for URIs.
       const unknownFrameAnn = {
+        $tag: 't3',
         id: 'ghi',
         uri: 'https://example.org',
       };
@@ -265,9 +239,7 @@ describe('FrameSyncService', () => {
         uri: iframeAnn.uri,
       });
 
-      fakeStore.setState({
-        annotations: [mainFrameAnn, iframeAnn, unknownFrameAnn],
-      });
+      fakeStore.addAnnotations([mainFrameAnn, iframeAnn, unknownFrameAnn]);
 
       // Check that expected annotations were sent to each frame.
       assert.calledWithMatch(
@@ -284,6 +256,7 @@ describe('FrameSyncService', () => {
 
     it('sends annotation to first frame if there is no frame with matching URL or main frame', async () => {
       const annotation = {
+        $tag: 't1',
         id: 'abc',
         uri: 'urn:book-id:1234',
       };
@@ -299,9 +272,7 @@ describe('FrameSyncService', () => {
         uri: 'https://publisher.com/books/1234/chapter1.html',
       });
 
-      fakeStore.setState({
-        annotations: [annotation],
-      });
+      fakeStore.addAnnotations([annotation]);
 
       assert.calledWithMatch(
         guestRPC().call,
@@ -316,14 +287,10 @@ describe('FrameSyncService', () => {
       emitGuestEvent('documentInfoChanged', frameInfo);
 
       const ann2 = Object.assign({}, fixtures.ann, { $tag: 't2', id: 'a2' });
-      fakeStore.setState({
-        annotations: [fixtures.ann],
-      });
+      fakeStore.addAnnotations([fixtures.ann]);
       guestRPC().call.reset();
 
-      fakeStore.setState({
-        annotations: [fixtures.ann, ann2],
-      });
+      fakeStore.addAnnotations([ann2]);
 
       assert.calledWithMatch(
         guestRPC().call,
@@ -335,9 +302,7 @@ describe('FrameSyncService', () => {
     it('does not send a "loadAnnotations" message for replies', async () => {
       await connectGuest();
 
-      fakeStore.setState({
-        annotations: [annotationFixtures.newReply()],
-      });
+      fakeStore.addAnnotations([annotationFixtures.newReply()]);
 
       assert.isFalse(guestRPC().call.calledWith('loadAnnotations'));
     });
@@ -351,11 +316,9 @@ describe('FrameSyncService', () => {
     it('sends a "publicAnnotationCountChanged" message to the frame when there are public annotations', async () => {
       await connectGuest();
       emitGuestEvent('documentInfoChanged', fixtures.htmlDocumentInfo);
-      fakeStore.frames()[0].isAnnotationFetchComplete = true;
+      fakeStore.updateFrameAnnotationFetchStatus(fixtures.htmlDocumentInfo.uri, true);
 
-      fakeStore.setState({
-        annotations: [annotationFixtures.publicAnnotation()],
-      });
+      fakeStore.addAnnotations([annotationFixtures.publicAnnotation()]);
 
       assert.calledWithMatch(
         hostRPC().call,
@@ -367,14 +330,12 @@ describe('FrameSyncService', () => {
     it('sends a "publicAnnotationCountChanged" message to the frame when there are only private annotations', async () => {
       await connectGuest();
       emitGuestEvent('documentInfoChanged', fixtures.htmlDocumentInfo);
-      fakeStore.frames()[0].isAnnotationFetchComplete = true;
+      fakeStore.updateFrameAnnotationFetchStatus(fixtures.htmlDocumentInfo.uri, true);
 
       const annot = annotationFixtures.defaultAnnotation();
       delete annot.permissions;
 
-      fakeStore.setState({
-        annotations: [annot],
-      });
+      fakeStore.addAnnotations([annot]);
 
       assert.calledWithMatch(
         hostRPC().call,
@@ -386,21 +347,17 @@ describe('FrameSyncService', () => {
     it('does not send a "publicAnnotationCountChanged" message to the frame if annotation fetch is not complete', async () => {
       await connectGuest();
       emitGuestEvent('documentInfoChanged', fixtures.htmlDocumentInfo);
-      fakeStore.frames()[0].isAnnotationFetchComplete = false;
 
       // Discard 'publicAnnotationCountChanged' message sent after frame connected.
       hostRPC().call.resetHistory();
 
-      fakeStore.setState({
-        annotations: [annotationFixtures.publicAnnotation()],
-      });
+      fakeStore.addAnnotations([annotationFixtures.publicAnnotation()]);
+
       assert.isFalse(hostRPC().call.calledWith('publicAnnotationCountChanged'));
     });
 
     it('does not send a "publicAnnotationCountChanged" message if there are no connected frames', () => {
-      fakeStore.setState({
-        annotations: [annotationFixtures.publicAnnotation()],
-      });
+      fakeStore.addAnnotations([annotationFixtures.publicAnnotation()]);
       assert.isFalse(hostRPC().call.calledWith('publicAnnotationCountChanged'));
     });
   });
@@ -412,13 +369,8 @@ describe('FrameSyncService', () => {
       frameSync.connect();
       await connectGuest();
 
-      fakeStore.setState({
-        annotations: [ann],
-      });
-
-      fakeStore.setState({
-        annotations: [],
-      });
+      fakeStore.addAnnotations([ann]);
+      fakeStore.removeAnnotations([ann]);
 
       assert.calledWithMatch(guestRPC().call, 'deleteAnnotation', ann.$tag);
     });
@@ -567,10 +519,6 @@ describe('FrameSyncService', () => {
           id: frameInfo.frameIdentifier,
           metadata: frameInfo.metadata,
           uri: frameInfo.uri,
-
-          // This would be false in the real application initially, but in these
-          // tests we pretend that the fetch completed immediately.
-          isAnnotationFetchComplete: true,
         },
       ]);
     });
```